File: ios/AIRFacebook/Functions/LogEventFunction.m
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
#import "LogEventFunction.h"
#import <Foundation/Foundation.h>
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import "AIRFacebook.h"
#import <FBSDKCoreKit/FBSDKCoreKit.h>

FREObject fb_logEvent( FREContext context, void *functionData, uint32_t argc, FREObject *argv ) {
    NSString* eventName = [MPFREObjectUtils getNSString:argv[0]];
    NSDictionary* params = (argv[1] == nil) ? nil : [MPFREObjectUtils getNSDictionary:argv[1]];
    double valueToSum = [MPFREObjectUtils getDouble:argv[2]];

    if( params != nil ) {
        if( valueToSum == 0.0 ) {
            [AIRFacebook log:[NSString stringWithFormat:@"Logging event %@ with params %@", eventName, params]];
            [FBSDKAppEvents logEvent:eventName parameters:params];
        } else {
            [AIRFacebook log:[NSString stringWithFormat:@"Logging event %@ with params %@ and valueToSum %f", eventName, params, valueToSum]];
            [FBSDKAppEvents logEvent:eventName valueToSum:valueToSum parameters:params];
        }
    } else {
        if( valueToSum == 0.0 ) {
            [AIRFacebook log:[NSString stringWithFormat:@"Logging event %@", eventName]];
            [FBSDKAppEvents logEvent:eventName];
        } else {
            [AIRFacebook log:[NSString stringWithFormat:@"Logging event %@ and valueToSum %f", eventName, valueToSum]];
            [FBSDKAppEvents logEvent:eventName valueToSum:valueToSum];
        }
    }

    [FBSDKAppEvents flush];

    return nil;
}

File: android/src/com/marpies/ane/facebook/functions/LogEventFunction.java
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.facebook.functions;

import android.os.Bundle;
import com.adobe.fre.FREArray;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.facebook.appevents.AppEventsLogger;
import com.marpies.ane.facebook.utils.AIR;
import com.marpies.ane.facebook.utils.FREObjectUtils;

public class LogEventFunction extends BaseFunction {

	private static AppEventsLogger mLogger = null;
	private static final double EPSILON = 0.0000001;

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		String eventName = FREObjectUtils.getStringFromFREObject( args[0] );
		Bundle parameters = null;
		if( args[1] != null ) {
			parameters = FREObjectUtils.getBundleFromFREArray( (FREArray) args[1] );
		}
		double valueToSum = FREObjectUtils.getDoubleFromFREObject( args[2] );

		if( parameters != null ) {
			if( isZero( valueToSum ) ) {
				getLogger().logEvent( eventName, parameters );
			} else {
				getLogger().logEvent( eventName, valueToSum, parameters );
			}
		} else {
			if( isZero( valueToSum ) ) {
				getLogger().logEvent( eventName );
			} else {
				getLogger().logEvent( eventName, valueToSum );
			}
		}

		AIR.log( "Logging event " + eventName + "." );

		return null;
	}

	public boolean isZero( double value ) {
		return value >= -EPSILON && value <= EPSILON;
	}

	private AppEventsLogger getLogger() {
		if( mLogger == null ) {
			mLogger = AppEventsLogger.newLogger( AIR.getContext().getActivity() );
		}
		return mLogger;
	}

}
